### PR TITLE
Document and test lack of inherited signaling in RBF policy

### PR DESCRIPTION
**Pull request by ariard**
Contrary to BIP125 or other full-node implementation (e.g btcd), Bitcoin Core's mempool policy doesn't implement inherited signaling. 

This PR documents our mempool behavior on this and add a test demonstrating the case.

### CODE REVIEW

**Comment by ariard** [comment]
Also, I had a look on the rest of RBF functional test coverage (i.e feature_rbf.py, mempool_accept.py, p2p_invalid_tx.py). AFAICT, the latest check on incremental relay fee increase doesn't seem to have coverage ?

**Comment by Sjors** [comment]
utACK fb2047b

File: test/functional/feature_rbf.py [hunk 1]
```diff
@@ -563,5 +566,68 @@ def test_rpc(self):
...
+            'vout': 0,
+            "sequence": 0xffffffff,
+        }], {self.nodes[0].getnewaddress(): Decimal("9.00000")})
+
+        replacement_child_tx = self.nodes[0].signrawtransactionwithwallet(replacement_child_tx)
+
+        # Broadcast replacement child tx
+        # BIP 125 :
+        # 1. The original transactions signal replaceability explicitly or through inheritance as described in the above
+        # Summary section.
```
Comment (Sjors) [inline test/functional/feature_rbf.py:626]: This is a good place to refer to the CVE so it's clear the behaviour here is _not_ following BIP125.

Comment (ariard) [reply test/functional/feature_rbf.py:626]: Yes already add a ref in new document, added another here if you think so.

**Comment by prayank23** [comment]
Concept ACK

Maybe we can mention this in [doc/bips.md](https://github.com/bitcoin/bitcoin/blob/master/doc/bips.md) as well

File: test/functional/feature_rbf.py (hunk 1 above)
Comment (Sjors) [reply test/functional/feature_rbf.py:626]: Probably a good idea; few people will read `validation.cpp` along side this test.

**Review by jonatack — COMMENTED** [review]
Concept ACK. Context/additional information:

- https://bitcoinops.org/en/newsletters/2021/05/12/#cve-2021-31876-discrepancy-between-bip125-and-bitcoin-core-implementation
- https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2021-May/018893.html

File: test/functional/feature_rbf.py [hunk 2]
```diff
@@ -563,5 +566,69 @@ def test_rpc(self):
         assert_equal(json0["vin"][0]["sequence"], 4294967293)
         assert_equal(json1["vin"][0]["sequence"], 4294967294)
 
+    def test_no_inherited_signaling(self):
+        # Send tx from which to conflict outputs later
+        base_txid = self.nodes[0].sendtoaddress(self.nodes[0].getnewaddress(), Decimal("10"))
+        self.nodes[0].generate(1)
+        self.sync_blocks()
+
+        # Create a explicitly opting parent transaction
```
Comment (jonatack) [inline test/functional/feature_rbf.py:575]:
```suggestion
        # Create an explicitly opt-in parent transactio
```

File: test/functional/feature_rbf.py [hunk 3]
```diff
@@ -563,5 +566,69 @@ def test_rpc(self):
...
+
+        replacement_parent_tx = self.nodes[0].signrawtransactionwithwallet(replacement_parent_tx)
+
+        # Test if parent tx can be replaced.
+        res = self.nodes[0].testmempoolaccept(rawtxs=[replacement_parent_tx['hex']], maxfeerate=0)[0]
+
+        # Parent can be replaced.
+        assert_equal(res['allowed'], True)
+
+        # Create a optout child tx spending opting parent
```
Comment (jonatack) [inline test/functional/feature_rbf.py:602]:
```suggestion
        # Create an opt-out child tx spending the opt-in parent
```

**Review by jonatack — COMMENTED** [review]
ACK 3d57b5eb053c7a34c3621875fa60401782a280b4

Happy to re-ACK if you update per the comments.

**Comment by ariard** [comment]
Thanks @jonatack and @Sjors, I think I've addressed all your comments, but lmk if you've more I'll address them quickly.

@prayank23 

Well in fact there is another well-known supplemental check implemented by our logic not mentioned in the BIP, which is the feerate comparison here : https://github.com/bitcoin/bitcoin/blob/ecf5f2c1a06edd8372402872525f8de1d4277453/src/validation.cpp#L848 and absolute fee checks is done on original transactions *and* their descendants, should we document all of them ?

**Comment by ariard** [comment]
@MarcoFalke I believe you're aware of few mempool-related RPCs which should be updated in consequence as follow-up of this PR?

**Comment by jonatack** [comment]
ACK 2eb0eeda39cab997f9a5f817f7c12e7fffeaf35a

**Comment by prayank23** [comment]
> Well in fact there is another well-known supplemental check implemented by our logic not mentioned in the BIP, which is the feerate comparison here :
> 
> https://github.com/bitcoin/bitcoin/blob/ecf5f2c1a06edd8372402872525f8de1d4277453/src/validation.cpp#L848
> 
> and absolute fee checks is done on original transactions _and_ their descendants, should we document all of them ?

Not sure about this. According to my understanding fee rate and fees both should be higher in replacement transaction.

Example:
Tx1 has only 1 input and two outputs (RBD enabled)
Tx2 is replacement transaction for Tx1 which has 5 inputs and 2 outputs.

It is possible that Tx2 uses a lower fee rate compared to Tx1 but more fees. Miners normally prioritize based on fee rate so a higher fee rate would be better.

However, if there is a difference in BIP and the implementation it should be mentioned in [doc/bip.md](https://github.com/bitcoin/bitcoin/blob/master/doc/bips.md) IMO

**Comment by benthecarman** [comment]
ACK 2eb0eeda39cab997f9a5f817f7c12e7fffeaf35a

Is a fix for the actual CVE being worked on as well?

**Comment by ariard** [comment]
@prayank23 

Yes, unless you have a strong opinion here, I prefer differences between BIPs and the implementation in another PR, like one adding test coverage for incremental relay fee.

@benthecarman 

Not really, as a fix, I'm currently studying moving towards full-rbf. If we do so, there wouldn't be any kind of signaling to proceed at all.

IMO, main motivation to not fix it is avoiding increasing the DoS surface of the mempool, where at replacement transaction submission you might have to traverse N * `package_ancestor_limits`. With N your number of allowed inputs for a `MAX_STANDARD_TX_WEIGHT-`-sized transaction.

**Comment by jonatack** [comment]
This has ACKs by @Sjors, @benthecarman and myself, if anyone would like to have a look.

**Comment by ajtowns** [comment]
> IMO, main motivation to not fix it is avoiding increasing the DoS surface of the mempool, where at replacement transaction submission you might have to traverse N * `package_ancestor_limits`. With N your number of allowed inputs for a `MAX_STANDARD_TX_WEIGHT`-sized transaction.

Adding a `bool rbf_enabled` to `CTxMemPoolEntry` that's true if the tx signals for rbf or any of its inputs have `rbf_enabled == true` would remove that problem fairly easily, as far as I can see.

**Comment by MarcoFalke** [comment]
> @MarcoFalke I believe you're aware of few mempool-related RPCs which should be updated in consequence as follow-up of this PR?

Sure, filed as #22209

**Comment by ariard** [comment]
@ajtowns

IIUC we would have to re-order RBF opt-in checks after fetching parents coins from the cache in `PreChecks`, but I don't believe such re-order would change anything in DoS prevention strategy around mempool acceptance. Any `CTxMemPoolEntry` would inherit `rbf_enable == true` if one of its parent entries have the flag sets so, that way you only check first-depth of ancestors of the replaced transactions.

At first sight, I don't see any blocker for this approach, though I was planning to propose soon moving towards full-rbf. Beyond simplifying a bit `PreChecks` code sanity, opt-in RBF raises issues for multi-party funded transactions. If full-rbf turns as a flames war like last time, we can consider again this approach to fix the discrepancy between the spec and our code.

**Pull request merged**